File: LJ_gas_run_MD.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
LJ_gas_run_MD.py

Main program for running molecular dynamics simulations using Lennard-Jones particles.
Initializes the system, runs the integrator loop, records energy and trajectory data, 
and visualizes results.

Author: Bettina Keller
Created: May 28, 2025

This script imports all classes and functions from md_simulation.py and controls
the simulation workflow.

"""

#----------------------------------------------------------------
#   I M P O R T S
#----------------------------------------------------------------
import numpy as np
from scipy.constants import R
import matplotlib.pyplot as plt

import time
from datetime import datetime

from LJ_gas import(
    ParticleSystem,
    SimulationParameters,
    simulate_NVE_step,
    simulate_NVT_step,
    initialize_positions,
    initialize_velocities,
    initialize_LJ,
    calculate_force,
    density,
    write_xyz_trajectory,
    potential_energy,
    kinetic_energy,
    instantaneous_temperature,
    ideal_gas_pressure,
    write_output_file
    )

#----------------------------------------------------------------
#   F U N C T I O N S
#----------------------------------------------------------------
# Define tic and toc functions
def tic():
    """Start a timer."""
    global _tic_time
    _tic_time = time.time()

def toc():
    """Stop the timer and return the elapsed time in seconds."""

    elapsed_time = None
    
    if '_tic_time' in globals():
        elapsed_time = time.time() - _tic_time
    
    else:
        print("Error: tic() was not called before toc()")
    
    return elapsed_time


#----------------------------------------------------------------
#   P A R A M E T E R S
#----------------------------------------------------------------
# system
n_particles = 200
mass_argon =  39.95             # mass in u = 1e-3 kg/mol
sigma_argon = 0.34              # sigma in nm     Argon: 0.34
epsilon_argon = 120*R*1e-3      # epsilon in kJ/mol Argon: 120

# simulation
dt = 0.1             # ps
n_steps = 1000 
temperature = 300     # K
box_length = 100      # nm
tau_thermostat = 1  # thermostat coupling constant in 1/ps
rij_min = 1e-2      # nm
NVT = True          # switch to decide between NVT and NVE

# output
# Make sure the appropriate subfolder exists in the working directory
file_name_base = "simulations/my_simulation"  # file name for all output files

#----------------------------------------------------------------
#   P R O G R A M
#----------------------------------------------------------------
# start the timer
tic()

#
# initialize simulation parameters
#
sim = SimulationParameters(dt = dt, 
                           n_steps = n_steps, 
                           temperature = temperature, 
                           box_length = box_length, 
                           tau_thermostat = tau_thermostat,
                           rij_min=rij_min,
                           is_NVT = NVT
                           )

#
# initialize ParticleSystem 
#
ps = ParticleSystem(n_particles)

# fill in the parameters for argon
for i in range(n_particles): 
    ps.set_parameters(i, type="Ar", mass=mass_argon, sigma=sigma_argon, epsilon=epsilon_argon)

# initialize LJ params
initialize_LJ(ps)

# set initial positions     
initialize_positions(ps, sim.box_length)

# set initial velocities     
initialize_velocities(ps, sim.temperature)

# calculate force according to initial positions
calculate_force(ps, sim)

# calculate initial values of variable properties
E_pot_init = potential_energy(ps, sim)
E_kin_init = kinetic_energy(ps)
T_init = instantaneous_temperature(ps)
P_init = ideal_gas_pressure(ps, sim)


# initialize position trajectory
position_trajectory = np.zeros((sim.n_steps+1, n_particles, 3))
position_trajectory[0,:,:] = ps.position # initial position

# initialize energy trajectory
energy_trajectory = np.zeros((sim.n_steps+1, 4))
energy_trajectory[0,0] = potential_energy( ps, sim)       # potential energy
energy_trajectory[0,1] = kinetic_energy(ps)               # kinetic energy
energy_trajectory[0,2] = instantaneous_temperature(ps)    # instantaneous temperature
energy_trajectory[0,3] = ideal_gas_pressure(ps, sim)      # ideal gas pressure


#--------------------------------------------------
#  The acutal MD simulation
#--------------------------------------------------
for i in range(sim.n_steps):
    if sim.is_NVT==True:
        simulate_NVT_step(ps, sim)
    else: 
        simulate_NVE_step(ps, sim)
        
    # store updated positions
    position_trajectory[i+1,:,:] = ps.position # store updated positions

    # store updated energies, temperature and pressure
    energy_trajectory[i+1,0] = potential_energy( ps, sim)     # potential energy
    energy_trajectory[i+1,1] = kinetic_energy(ps)             # kinetic energy
    energy_trajectory[i+1,2] = instantaneous_temperature(ps)  # instantaneous temperature
    energy_trajectory[i+1,3] = ideal_gas_pressure(ps, sim)    # ideal gas pressure


#--------------------------------------
# W R I T E    T R A J E C T O R I E S 
#--------------------------------------
elapsed_time = toc()   # stop the timer

# write position trajectory to file
write_xyz_trajectory(file_name_base + "_pos.xyz", position_trajectory, atom_symbols=ps.type)

# write output file 
output = write_output_file(file_name_base + ".out", ps, sim, elapsed_time)
print(output)

# write energy trajectory to file (binary and text)
np.save(file_name_base + "_ene.npy", energy_trajectory)
np.savetxt(file_name_base + "_ene.dat", energy_trajectory, fmt="%.6e", header="#E_pot  E_kin  T  P", comments='')


#----------------------------------------------------
# P L O T   E N E R G Y   T R A J E C T O R I E S
#----------------------------------------------------
# set time axis
time_ps = np.arange(sim.n_steps + 1) * sim.dt

#
# potential energy
# 
E_pot_min = np.mean(energy_trajectory[:,0]) - 1   # lower limit of E_pot axis
E_pot_max = np.mean(energy_trajectory[:,0]) + 1   # upper limit of E_pot axis 

plt.figure(figsize=(8, 6))
plt.plot(time_ps, energy_trajectory[:,0]) 
plt.ylim(E_pot_min, E_pot_max)
plt.xlabel("time [ps]", fontsize=14)
plt.ylabel("E_pot [kJ/mol]", fontsize=14)

plt.savefig(file_name_base + "_Epot.png", dpi=300, bbox_inches='tight')
plt.show()

#
# kinetic energy
# 
E_kin_min = np.mean(energy_trajectory[:,1]) - 100   # lower limit of E_kin axis
E_kin_max = np.mean(energy_trajectory[:,1]) + 100   # upper limit of E_kin axis 

plt.figure(figsize=(8, 6))
plt.plot(time_ps, energy_trajectory[:,1]) 
plt.ylim(E_kin_min, E_kin_max)
plt.xlabel("time [ps]", fontsize=14)
plt.ylabel("E_kin [kJ/mol]", fontsize=14)

plt.savefig(file_name_base + "_Ekin.png", dpi=300, bbox_inches='tight')
plt.show()

#
# temperature
# 
T_min = np.mean(energy_trajectory[:,2]) - 100   # lower limit of T axis
T_max = np.mean(energy_trajectory[:,2]) + 100   # upper limit of T axis 

plt.figure(figsize=(8, 6))
plt.plot(time_ps, energy_trajectory[:,2]) 
plt.ylim(T_min, T_max)
plt.xlabel("time [ps]", fontsize=14)
plt.ylabel("T [K]", fontsize=14)

plt.savefig(file_name_base + "_T.png", dpi=300, bbox_inches='tight')
plt.show()

#
# pressure
# 
P_min = np.mean(energy_trajectory[:,3]) - 200   # lower limit of P axis
P_max = np.mean(energy_trajectory[:,3]) + 200   # upper limit of P axis 

plt.figure(figsize=(8, 6))
plt.plot(time_ps, energy_trajectory[:,3]) 
plt.ylim(P_min, P_max)
plt.xlabel("time [ps]", fontsize=14)
plt.ylabel("P [Pa]", fontsize=14)

plt.savefig(file_name_base + "_P.png", dpi=300, bbox_inches='tight')
plt.show()

File: LJ_gas_run_MD.jl
include("LJ_gas.jl")

using AMDGPU
using ProgressMeter
using LinearAlgebra

# system
n_particles = 2000

# LENNARD JONES PARAMS: https://doi.org/10.1080/00268976.2016.1246760
#type_A = "Kr"
#mass_A =  83.798             # mass in u = 1e-3 kg/mol
#sigma_A = 0.3627              # sigma in nm     
#epsilon_A = 162.58 * R * 1e-3     # epsilon in kJ/mol

type_A = "Ne"
mass_A =  20.18             # mass in u = 1e-3 kg/mol
sigma_A = 0.2801              # sigma in nm     
epsilon_A = 33.921 * R * 1e-3     # epsilon in kJ/mol

type_B = "Ar"
mass_B =  39.948             # mass in u = 1e-3 kg/mol
sigma_B = 0.33952              # sigma in nm     
epsilon_B = 116.79 * R * 1e-3     # epsilon in kJ/mol

#type_B = "Rn"
#mass_B =  222             # mass in u = 1e-3 kg/mol 
#sigma_B= 0.417           # sigma in nm     
#epsilon_B = 300 * R * 1e-3   # epsilon in kJ/mol 

#type_B = "He"
#mass_B =  4.002602             # mass in u = 1e-3 kg/mol 
#sigma_B= 0.228           # sigma in nm     
#epsilon_B = 10.2 * R * 1e-3  # epsilon in kJ/mol 

# simulation
dt = 0.02          # ps
n_steps = 50000
temperature = 300     # K
end_temperature = 0
box_length = 30      # nm
tau_thermostat = 1  # thermostat coupling constant in 1/ps
rij_min = 1e-2      # nm
rij_min_starting = 0.3 # nm
NVT = true          # switch to decide between NVT and NVE
equilibrate = false # turn on NVT equilibration for all simulations

# Metadata 
save_frequency = 10 # save every n-th step
filename_base = "simulations/NeAr_Large_NVT_300"

# Switch between CPU and GPU computation by choosing the array type used for initialization
ARRAYTPE = ROCArray{Float32} # GPU
#ARRAYTPE = Array{Float32} # CPU single precision
#ARRAYTPE = Array{Float64} # CPU double precision

#calculate save points
n_saves = div(n_steps, save_frequency) + 1

sim = SimulationParameters(dt, n_steps, temperature, box_length, tau_thermostat, rij_min, Inf, NVT, save_frequency)

# initialize ParticleSystem 
#
ps = ParticleSystem(n_particles, ARRAYTPE)

# Split systn_particles in two parts
n_A = div(n_particles, 2)
n_B = n_particles - n_A

box_A = ((0, box_length / 2), (0, box_length), (0, box_length))
box_B = ((box_length / 2, box_length), (0, box_length), (0, box_length))


# fill in the parameters

ps.type[1:n_A] = fill(type_A,n_A)
ps.mass[1:n_A] = fill(mass_A,n_A)
ps.sigma[1:n_A]= fill(sigma_A,n_A)
ps.epsilon[1:n_A] = fill(epsilon_A,n_A)


ps.type[n_A+1:end] = fill(type_B,n_B)
ps.mass[n_A+1:end] = fill(mass_B,n_B)
ps.sigma[n_A+1:end] = fill(sigma_B,n_B)
ps.epsilon[n_A+1:end] = fill(epsilon_B,n_B)

function get_position(xrange, yrange, zrange)
    pos = zeros(3)
    pos[1] = rand(Uniform(xrange...))
    pos[2] = rand(Uniform(yrange...))
    pos[3] = rand(Uniform(zrange...))
    return pos
end

# initialize LJ params
initialize_LJ!(ps)

# guess initial position array
pos_initial = zeros(3, n_particles)

for i in 1:n_A
    pos_initial[:, i] = get_position(box_A...)
end

for i in n_A+1:n_particles
    pos_initial[:, i] = get_position(box_B...)
end

# remove distances which are to small
for i in 1:100
    r = [p1[i] - p2[i] for i in 1:3, p1 in eachslice(pos_initial, dims=2), p2 in eachslice(pos_initial, dims=2)]
    # Apply periodic boundary conditions
    @. r = r - round(r / box_length) * box_length

    dist = reshape(norm.(eachslice(r,dims=(2, 3))), n_particles^2)

    # set diagonal elements to inf
    for i in 1:n_particles
        dist[LinearIndices((n_particles, n_particles))[i,i]] = Inf
    end

    cutoff = findfirst(sort(dist) .<= rij_min_starting)

    if !isnothing(cutoff)
        for i in sortperm(dist)[1:cutoff]
            index = CartesianIndices((n_particles, n_particles))[i][1]
            if index > n_A
                pos_initial[:,index] = get_position(box_B...)
            else
                pos_initial[:,index] = get_position(box_A...)
            end
        end
    else
        println("Placed particles respecting rij_min_starting. Current minimum distance: ", minimum(dist))
        break
    end
    
    i == 100 && println("Could not place all particles with rij_min_starting. Current minimum distance: ", minimum(dist))
end

# set initial positions     
ps.position[:, :] = pos_initial


# set initial velocities     
initialize_velocities!(ps, sim.temperature)

# calculate force according to initial positions
calculate_force!(ps, sim)

# calculate initial values of variable properties
E_pot_init = potential_energy(ps, sim)
E_kin_init = kinetic_energy(ps)
T_init = instantaneous_temperature(ps)
P_init = ideal_gas_pressure(ps, sim)

position_trajectory = similar(ps.position, (3, n_particles, n_saves))
position_trajectory[:,:,1] = ps.position

# initialize energy trajectory
energy_trajectory = zeros((4,n_saves))
energy_trajectory[1,1] = potential_energy(ps, sim)       # potential energy
energy_trajectory[2,1] = kinetic_energy(ps)               # kinetic energy
energy_trajectory[3,1] = instantaneous_temperature(ps)    # instantaneous temperature
energy_trajectory[4,1] = ideal_gas_pressure(ps, sim)      # ideal gas pressure

# Equilibrate by simulating a few NVT steps
if equilibrate
    sim_equil = SimulationParameters(dt/500, n_steps, temperature, box_length, tau_thermostat, rij_min, Inf, NVT, save_frequency)

    for k in 1:1000
        simulate_NVT_step!(ps, sim_equil)
        if instantaneous_temperature(ps) < sim_equil.temperature * 1.1
            println("Rough equilibration finished after $k steps.")
            break
        end
    end

    for k in 1:50
        simulate_NVT_step!(ps, sim_equil)
    end

    println("Equilibration done with current temperature: ", instantaneous_temperature(ps))
end


p = Progress(sim.n_steps, dt = 1, showspeed=true)
generate_showvalues(iter, current_save_pos) = () -> [(:iter,iter), ("Potential Energy", energy_trajectory[1,current_save_pos+1]), ("Temperature", energy_trajectory[3,current_save_pos+1])]

stats = @timed for i in 1:n_steps
    if NVT
        simulate_NVT_step!(ps, sim)
    else
        simulate_NVE_step!(ps, sim)
    end

    current_save_pos = div(i,save_frequency)
    if mod(i,save_frequency) == 0
        # store updated positions
        position_trajectory[:,:,current_save_pos+1] = ps.position # store updated positions

        # store updated energies, temperature and pressure
        energy_trajectory[1,current_save_pos+1] = potential_energy( ps, sim)     # potential energy
        energy_trajectory[2,current_save_pos+1] = kinetic_energy(ps)             # kinetic energy
        energy_trajectory[3,current_save_pos+1] = instantaneous_temperature(ps)  # instantaneous temperature
        energy_trajectory[4,current_save_pos+1] = ideal_gas_pressure(ps, sim)    # ideal gas pressure

        if !iszero(end_temperature)
            @assert end_temperature <= temperature
            T = end_temperature + (temperature - end_temperature) * (1 - div(i,save_frequency) / div(n_steps,save_frequency))
            global sim = SimulationParameters(dt, n_steps, T, box_length, tau_thermostat, rij_min, Inf, NVT, save_frequency)
        end

    end
    next!(p, showvalues = generate_showvalues(i, current_save_pos))
end

energies = energy_trajectory[1,:] .+ energy_trajectory[2,:]
temperatures = energy_trajectory[3,:]
pressures = energy_trajectory[3,:]


write_xyz_trajectory(filename_base, ps, position_trajectory)
write_energy_trajectory(filename_base, energy_trajectory)
write_output_file(filename_base,ps,sim,stats.time)

using GLMakie
lines(energies)
